### PR TITLE
Support nested arrays for `min`/`max` aggregations in groupby and reduction

### PR DESCRIPTION
**Pull request by ttnghia**
This adds `ARRAY` type support for  `min` and `max` aggregations, which is simply an adoption of the changes from cudf.
In addition,  it also reverts the changes in https://github.com/NVIDIA/spark-rapids/pull/8703 since the issue has been fixed.

Depends on:
 * https://github.com/rapidsai/cudf/pull/13676

Closes https://github.com/NVIDIA/spark-rapids/issues/4929 and closes https://github.com/NVIDIA/spark-rapids/issues/8668.

Contribute to https://github.com/NVIDIA/spark-rapids/issues/8550.

### CODE REVIEW

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuOverrides.scala [hunk 1]
```diff
@@ -2110,12 +2110,14 @@ object GpuOverrides extends Logging {
       "Max aggregate operator",
       ExprChecksImpl(
         ExprChecks.reductionAndGroupByAgg(
-          (TypeSig.commonCudfTypes + TypeSig.DECIMAL_128 + TypeSig.NULL + TypeSig.STRUCT)
+          (TypeSig.commonCudfTypes + TypeSig.DECIMAL_128 + TypeSig.NULL + TypeSig.STRUCT +
+            TypeSig.ARRAY)
             .nested(TypeSig.commonCudfTypes + TypeSig.DECIMAL_128 + TypeSig.NULL +
```
Comment (revans2) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuOverrides.scala:2115]: nit:

```
(TypeSig.commonCudfTypes + TypeSig.DECIMAL_128 + TypeSig.NULL + TypeSig.STRUCT +
            TypeSig.ARRAY).nested()
```

If no data types are passed into nested it assumes it is the same data types at the top level.

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuOverrides.scala [hunk 2]
```diff
@@ -2110,12 +2110,14 @@ object GpuOverrides extends Logging {
...
               TypeSig.STRUCT + TypeSig.ARRAY),
           TypeSig.orderable,
           Seq(ParamCheck("input",
-            (TypeSig.commonCudfTypes + TypeSig.DECIMAL_128 + TypeSig.NULL + TypeSig.STRUCT)
+            (TypeSig.commonCudfTypes + TypeSig.DECIMAL_128 + TypeSig.NULL + TypeSig.STRUCT +
+              TypeSig.ARRAY)
```
Comment (revans2) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuOverrides.scala:2120]: same nit here.

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuOverrides.scala [hunk 3]
```diff
@@ -2137,12 +2139,14 @@ object GpuOverrides extends Logging {
       "Min aggregate operator",
       ExprChecksImpl(
         ExprChecks.reductionAndGroupByAgg(
-          (TypeSig.commonCudfTypes + TypeSig.DECIMAL_128 + TypeSig.NULL + TypeSig.STRUCT)
+          (TypeSig.commonCudfTypes + TypeSig.DECIMAL_128 + TypeSig.NULL + TypeSig.STRUCT +
```
Comment (revans2) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuOverrides.scala:2142]: and here.

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuOverrides.scala [hunk 4]
```diff
@@ -2137,12 +2139,14 @@ object GpuOverrides extends Logging {
...
+              TypeSig.ARRAY)
             .nested(TypeSig.commonCudfTypes + TypeSig.DECIMAL_128 + TypeSig.NULL +
               TypeSig.STRUCT + TypeSig.ARRAY),
           TypeSig.orderable,
           Seq(ParamCheck("input",
-            (TypeSig.commonCudfTypes + TypeSig.DECIMAL_128 + TypeSig.NULL + TypeSig.STRUCT)
+            (TypeSig.commonCudfTypes + TypeSig.DECIMAL_128 + TypeSig.NULL + TypeSig.STRUCT +
```
Comment (revans2) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuOverrides.scala:2148]: and here.

File: integration_tests/src/main/python/hash_aggregate_test.py [hunk 5]
```diff
@@ -1844,11 +1844,9 @@ def test_std_variance_partial_replace_fallback(data_gen,
...
 @ignore_order(local=True)
 @pytest.mark.parametrize('data_gen',  gens_for_max_min, ids=idfn)
-def test_min_max_for_single_level_struct(data_gen):
+def test_min_max_for_nested_types(data_gen):
     df_gen = [
-        ('a', StructGen([
-                ('aa', data_gen),
-                ('ab', data_gen)])),
+        ('a', data_gen),
         ('b', RepeatSeqGen(IntegerGen(), length=20))]
```
Comment (revans2) [inline integration_tests/src/main/python/hash_aggregate_test.py:1851]: You added in support for a reduction on min/max (no group by keys).  Could we also add in a test for that case too?

Comment (ttnghia) [reply integration_tests/src/main/python/hash_aggregate_test.py:1851]: Yes this test function already has both reduction and groupby tests.

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuOverrides.scala (hunk 1 above)
Comment (ttnghia) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuOverrides.scala:2115]: Sorry I'm not clear about what to do. Can you show change suggestions, please?

Comment (revans2) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuOverrides.scala:2115]: No What I am suggesting is exactly what I wrote in the comment.

```
(TypeSig.commonCudfTypes + TypeSig.DECIMAL_128 + TypeSig.NULL + TypeSig.STRUCT +
            TypeSig.ARRAY).nested()
```

It is equivalent to the longer code
```
(TypeSig.commonCudfTypes + TypeSig.DECIMAL_128 + TypeSig.NULL + TypeSig.STRUCT +
            TypeSig.ARRAY).nested(TypeSig.commonCudfTypes + TypeSig.DECIMAL_128 + TypeSig.NULL + TypeSig.STRUCT +
            TypeSig.ARRAY)
```

**Review by revans2 — DISMISSED** [review]

**Comment by ttnghia** [comment]
build

**Comment by ttnghia** [comment]
build

**Comment by ttnghia** [comment]
Just realize that this may have conflicts with with the just-merged PR (https://github.com/NVIDIA/spark-rapids/pull/7465). Will manually merge later and build again.

**Comment by ttnghia** [comment]
Now it also depends on https://github.com/rapidsai/cudf/pull/13701.

**Review by jlowe — APPROVED** [review]

**Pull request merged**